### DISCUSSION TITLE
Executable build with NimGL cannot load cimgui.dylib

### DISCUSSION THREAD

**Discussion by memononen** [cavariux/nimgl#61] [Q&A]
Hi!

I'm new to Nim and NimGL. I created a new project using `nimble init`, added `requires "nimgl >= 1.0.0"` to the project's .nimble file, and copied the `timgui.nim` to replace the hello world .nim file created by nimble.

I can see that it retrieves and builds the library, and nim builds my project fine too. But I get following error when trying to run the created executable (hello world worked just fine):
```
could not load: cimgui.dylib
(compile with -d:nimDebugDlOpen for more information)
```
Is there anything obviously stupid I'm doing?

I have `~/.nimble/bin` added $PATH.

  - Version: 1.1.6
  - OS: MacOS
  - Bindings: OpenGL/GLFW/ImGui
  - Backend: C

**Comment by lmariscal** [comment]
The issue here is that ImGui is C++ while you are using the C backend. NimGL is making use of C bindings (cimgui) which needs a DLL in order to access the C++ code from a C interface.

https://github.com/nimgl/nimgl/discussions/58#discussioncomment-226225

You can follow those instructions to build the DLL from the NimGL's fork of cimgui. If you are in Mac the default generator will be a XCode project, you can change the generator with the `-G` flag in the cmake step to compile with a Makefile.

If you want to skip creating a DLL to load the definitions, you can use the C++ backend which automatically compile cimgui and imgui and include both statically in your executable.

**Reply by memononen** [reply]
Thanks for the swift reply! I added `backend="cpp"` into my projects .nimble file, and I got following error:

```
error: invalid argument '-std=gnu++14' not allowed with 'Objective-C'
Error: execution of an external compiler program 'clang++ -c -std=gnu++14 -funsigned-char  -I/Users/mikko/.nimble/pkgs/nimgl-1.1.6/nimgl/private/cimgui -DIMGUI_DISABLE_OBSOLETE_FUNCTIONS=1 -D_GLFW_COCOA -D_GLFW_USE_CHDIR -D_GLFW_USE_MENUBAR -D_GLFW_USE_RETINA -DGLFW_EXPOSE_NATIVE_COCOA -DGLFW_EXPOSE_NATIVE_NSGL   -I/Users/mikko/.choosenim/toolchains/nim-1.4.2/lib -I/Users/mikko/code/nim_crowd/src -o /Users/mikko/.cache/nim/nim_crowd_d/cocoa_init.m.o /Users/mikko/.nimble/pkgs/nimgl-1.1.6/nimgl/private/glfw/src/cocoa_init.m' failed with exit code: 1


       Tip: 4 messages have been suppressed, use --verbose to show them.
     Error: Build failed for package: nim_crowd
        ... Execution failed with exit code 256
        ... Command: /Users/mikko/.nimble/bin/nim cpp --colors:on --noNimblePath -d:NimblePkgVersion=0.1.0 --path:/Users/mikko/.nimble/pkgs/nimgl-1.1.6 --hints:off -o:/Users/mikko/code/nim_crowd/nim_crowd /Users/mikko/code/nim_crowd/src/nim_crowd.nim
        ```

**Reply by memononen** [reply]
Building the `cimgui.dylib` manually (as per above link) and using C backend works, though.
